pkg/conversion/conversion_rules.go

```diff
@@ -0,0 +1,241 @@
+	}
+
+	program, err := env.Program(ast,
+		cel.EvalOptions(cel.OptOptimize),
+		cel.OptimizeRegex(library.ExtensionLibRegexOptimizations...),
+		cel.InterruptCheckFrequency(celCheckFrequency),
+	)
+	if err != nil {
+		return nil, field.Invalid(rulePath.Child("transformation"), rule.Transformation, fmt.Sprintf("error creating CEL program: %v", err))
+	}
+
+	cr.Program = program
+
+	return cr, nil
+}
+
+func createCELEnv(structuralSchema *structuralschema.Structural) (*cel.Env, error) {
+	env, err := cel.NewEnv(cel.HomogeneousAggregateLiterals())
+	if err != nil {
+		return nil, fmt.Errorf("error creating CEL environment: %w", err)
+	}
+
+	registry := model.NewRegistry(env)
+
+	// inline local copy of upstream's generateUniqueSelfTypeName()
+	scopedTypeName := fmt.Sprintf("selfType%d", time.Now().Nanosecond())
+
+	ruleTypes, err := model.NewRuleTypes(scopedTypeName, structuralSchema, true, registry)
+	if err != nil {
+		return nil, fmt.Errorf("error creating rule types: %w", err)
+	}
+	if ruleTypes == nil {
+		return nil, fmt.Errorf("unexpected nil rule types")
+	}
+
+	opts, err := ruleTypes.EnvOptions(env.TypeProvider())
+	if err != nil {
+		return nil, fmt.Errorf("error getting CEL environment options: %w", err)
+	}
+
+	root, ok := ruleTypes.FindDeclType(scopedTypeName)
+	if !ok {
+		rootDecl := model.SchemaDeclType(structuralSchema, true)
+		if rootDecl == nil {
+			return nil, fmt.Errorf("unable to find CEL decl type for %s", structuralSchema.Type)
+		}
+		root = rootDecl.MaybeAssignTypeName(scopedTypeName)
+	}
+
+	var propDecls []*expr.Decl
+	propDecls = append(propDecls, decls.NewVar("self", root.ExprType()))
+
+	opts = append(opts, cel.Declarations(propDecls...), cel.HomogeneousAggregateLiterals())
+	opts = append(opts, library.ExtensionLibs...)
+	return env.Extend(opts...)
+}
+
+func getStructuralSchemaForField(s *structuralschema.Structural, fields ...string) (*structuralschema.Structural, error) {
+	// Start with the passed in schema
+	cursor := s
+
+	// Keep track of which fields we've already visited so we can be specific in our errors
+	visited := make([]string, 0, len(fields))
+
+	// Calculate the index of the target field's parent. For example, if we're looking up the schema for
+	// spec.name.first, the parent of "first" is "name", and its index is 1.
+	//
+	// spec.name.first ==> 1
+	//  0    1     2
+	lastParentIndex := len(fields) - 2
+
+	// Starting with the initial field (e.g. "spec"), try to resolve the next segment (e.g. "name") until we get
+	// to the desired field (e.g. "first").
+	for i, f := range fields {
+		// Verify that each intermediate field is an object. While iterating, i-1 is the index of the parent field.
+		if lastParentIndex >= 0 && (i-1) <= lastParentIndex && cursor.Type != "object" {
+			return nil, fmt.Errorf("expected field %q to be an object", strings.Join(visited, "."))
+		}
```

Comment:
Assuming I've understood this logic correctly, these checks are basically saying "every field in the path before the last one has to be of type `object`. If that's the case, given that we're iterating through `fields` from outermost to innermost, couldn't we check the `Type` field before doing the `continue` in Properties/AdditionalProperties below?